File: Aiyagari1994/Python/Functions/fnHHobj.py
import numpy as np

def fnHHobj(
        x: float, 
        beta: float, 
        budget: float, 
        grida: np.ndarray, 
        Na: float, 
        expvf: np.ndarray
        ) -> float:
    """
    Objective Function With Interpolation for Wealth

    Args:
        x: Value of aprime (a') chosen wealth
        beta: Discount factor
        budget: Budget constraint
        grida: Wealth grid
        Na: Number of grid points for wealth grid 
        expvf: Expected value of value function defined over wealth grid

    Returns:
        float: Value of the objective function
    """      
    aprime = x.item() if np.ndim(x) > 0 else x 
    c = np.maximum(1e-8, budget - aprime)

    aLow = np.searchsorted(grida, aprime, side='right') - 1 
    aLow = np.maximum(0, np.minimum(Na - 2, aLow))
    aHigh = aLow + 1

    wtLow = (grida[aHigh] - aprime) / (grida[aHigh] - grida[aLow])
    wtLow = np.clip(wtLow, 0, 1)
    wtHigh = 1 - wtLow  

    expval = wtLow*expvf[aLow] + wtHigh*expvf[aHigh]

    val = -(np.log(c) + beta * expval)
    return val

File: Aiyagari1994/Matlab/Functions/fnHHobj.m
%% Household Objective Funciton - Pass into a Solver/Minimizer %%%%%%%%%%%%
function val = fnHHobj(x, beta, budget, grid, Ngrid, expvf)
%   Objective Function with Interpolation 
%
%   Args:
%       x: value to evaluate function at 
%       beta: discount factor 
%       budget: total resources avaiable (upper constraint)
%       grid: grid to interpolate x on 
%       Ngrid: number of gridpoints in grid 
%       expvf: expected value of value function (agent's continuation
%       value)
%
%   Returns:
%       val: scalar value
%
    aprime = x;
    c = max(1e-8, budget - aprime);
    [LB, UB, wtLB, wtUB] = fnInterp1dGrid(aprime, grid, Ngrid);
    Evalue = wtLB*expvf(LB) + wtUB*expvf(UB);
    val = -(log(c) + beta * Evalue);
end